jax/interpreters/batching.py

```diff
@@ -123,22 +123,36 @@ def sublift(self, val):
     return BatchTracer(self, val.val, val.batch_dim)
 
   def get_primitive_batcher(self, primitive):
-    if primitive in initial_style_batchers:
-      return partial(initial_style_batchers[primitive],
-                    axis_name=self.axis_name,
-                    main_type=self.main.trace_type)
     try:
       return primitive_batchers[primitive]
     except KeyError as err:
       msg = "Batching rule for '{}' not implemented"
       raise NotImplementedError(msg.format(primitive)) from err
 
+  def get_axis_primitive_batcher(self, primitive, frame):
+    return partial(axis_primitive_batchers[primitive],
+        frame.size, frame.name, frame.main_trace.trace_type)
+
+  def get_frame(self, vals, dims) -> core.AxisEnvFrame:
+    if self.axis_name is core.no_axis_name:
+      # If axis name is `no_axis_name` we can't find it via `core.axis_name` so we
+      # reconstruct it from the information we have available
+      axis_sizes = {x.shape[d] for x, d in zip(vals, dims) if d is not not_mapped}
+      assert len(axis_sizes) == 1
+      axis_size, = axis_sizes
+      return core.AxisEnvFrame(self.axis_name, axis_size, self.main)
+    return core.axis_frame(self.axis_name)
+
   def process_primitive(self, primitive, tracers, params):
     vals_in, dims_in = unzip2((t.val, t.batch_dim) for t in tracers)
-    if (primitive in collective_rules and
-          _main_trace_for_axis_names(self.main, core.used_axis_names(primitive, params))):
-      frame = core.axis_frame(self.axis_name)
-      val_out, dim_out = collective_rules[primitive](frame, vals_in, dims_in, **params)
+    is_axis_primitive = primitive in axis_primitive_batchers
+    is_initial_style = is_axis_primitive and primitive not in primitive_batchers
```

Comment:
I don't think this is correct (it might be right now, but it doesn't look like a generalizable principle). Why do we need it? It's not like we have to actually call the initial style batching rules if there are no batched inputs _and_ the axis is not used.

Reply:
It seems like although "collectives" and "initial style primitives" can share the same batching rule, there are still some differences between how they operate. 

1. collectives - have both standard batchers and axis batchers. Axis batcher is triggered when the frame name matches the axis name in the params. If there are batch dims, standard batcher is triggered and otherwise short circuit.
2. initial style - always trigger axis batcher if there are batch dims, short circuit otherwise (there is no ordinary batching rule so we have to avoid taking that code path).

The solutions I see are either:
1. combine standard and axis batchers into `get_primitive_batcher`. This involves passing `frame` into `get_primitive_batcher`.
2. Check if the primitive has both a `standard_batcher` and an `axis_primitive_batcher`. If it does, it's a collective otherwise it's an initial style primitive. That is the current solution.

I've added in the short-circuiting condition to address the issue you pointed out, but as you said, I'm not sure how to make this into a generalizable principle while there is still this difference between collectives and initial style primities.

Reply:
I see the problem now. Would the following workflow work?

1. If the primitive has an axis_primitive_batcher *and* mentions the name then call the axis primitive rule. 
2. If there are no BatchTracers in inputs, short circuit
3. Call the primitive_batcher if available, falling back to an axis_primitive_batcher if unavailable

Reply:
Right so this is the previous solution I implemented. The main issue is the 3rd point; we need the frame to fall back to the `axis_primitive_batcher` so I had previously augmented the `get_primitive_batcher` method to take in the `frame`. Do you think there's a better option than that?

Reply:
I think it's ok for `get_primitive_batcher` to take in `frame` if it's also supposed to handle `axis_primitive_batchers`.

Reply:
Done.